### ISSUE TITLE
Regression: localized MSVC linker progress messages are emitted as linker stdout warnings

### ISSUE THREAD

**Issue by gifnksm** [rust-lang/rust#159133]
### Code

I tried this code:

`Cargo.toml`

```toml
[package]
name = "foo"
version = "0.1.0"
edition = "2024"

[lib]
crate-type = ["cdylib"]
```

`src/lib.rs` (empty)

Build with:

```console
cargo build
```

I expected to see this happen:

The build should complete without any warnings.

Instead, this happened:

```text
warning: linker stdout: ライブラリ C:\Users\...\foo.dll.lib とオブジェクト C:\Users\...\foo.dll.exp を作成中
  |
  = note: `#[warn(linker_messages)]` on by default
```

The warning is emitted on every build.

According to the [Rust 1.97 release notes], common linker messages that have been diagnosed as false positives or intentional behavior are filtered out by rustc.

The linker message:

```text
ライブラリ ... とオブジェクト ... を作成中
```

translates to:

```text
Creating library ... and object ...
```

This appears to be the localized equivalent of one of those common linker messages.

[Rust 1.97 release notes]: https://blog.rust-lang.org/2026/07/09/Rust-1.97.0/#linker-output-no-longer-hidden-by-default

### Version it worked on

It most recently worked on: 1.96.0

### Version with regression

`rustc --version --verbose`:

```
rustc 1.97.0 (2d8144b78 2026-07-07)
binary: rustc
commit-hash: 2d8144b7880597b6e6d3dfd63a9a9efae3f533d3
commit-date: 2026-07-07
host: x86_64-pc-windows-msvc
release: 1.97.0
LLVM version: 22.1.6
```

`rustc +nightly --version --verbose`:

```
rustc 1.99.0-nightly (375b1431b 2026-07-10)
binary: rustc
commit-hash: 375b1431b7d89d1c2e2bc168c011848ae12b7d14
commit-date: 2026-07-10
host: x86_64-pc-windows-msvc
release: 1.99.0-nightly
LLVM version: 22.1.8
```

### Backtrace

N/A (no crash)

### Additional information

Environment:

- Windows 11 (display language: Japanese)
- Target: `x86_64-pc-windows-msvc`
- Visual Studio Build Tools 2022
- `link.exe` version 14.44.35228.0

The issue reproduces consistently with the minimal example above.

The linker is localized. For example, invoking `link.exe` without arguments displays Japanese help text.

Looking at [`compiler/rustc_codegen_ssa/src/back/link.rs`][rustc-src], the linker message filter appears to recognize English messages such as `"Creating library"`, but not the localized equivalent.

[rustc-src]: https://github.com/rust-lang/rust/blob/3b58636b30eb364ac72aeaf03d46347084ed87d1/compiler/rustc_codegen_ssa/src/back/link.rs#L913-L935

@rustbot modify labels: +regression-from-stable-to-stable -regression-untriaged

**Comment by ChrisDenton** [comment]
We do also attempt to force English output https://github.com/rust-lang/rust/blob/3b58636b30eb364ac72aeaf03d46347084ed87d1/compiler/rustc_codegen_ssa/src/back/linker.rs#L43

However, that only works if the English language pack is installed.

**Comment by gifnksm** [comment]
Additional finding: I believe the mechanism is now clear.

In my environment, the issue is caused by `link.exe` emitting localized (Japanese) progress text.
rustc sets `VSLANG=1033` when invoking `link.exe`, but this only has effect if the English language resources are installed in Visual Studio Build Tools.

What I observed:

- Without English language pack:
  - `link.exe` output remains Japanese
  - rustc reports the localized progress line as `warning: linker stdout`
- After installing English language pack:
  - `VSLANG=1033` takes effect
  - `link.exe` output becomes English
  - warning disappears (intended behavior)

So this issue seems to depend on whether English language resources are present, not just on `VSLANG` being set.

Question for maintainers: should rustc treat “no English language pack installed” as supported and still suppress this known-benign message, or is requiring English linker resources acceptable?

**Comment by ds84182** [comment]
Are the language resources in a known format? If some Win32 API can read them, rustc could use that as a fallback when the English language resources aren't there.

**Comment by workingjubilee** [comment]
Are you hoping there is a file somewhere we can use for a symbol-based lookup for the messages we want to filter?

**Comment by workingjubilee** [comment]
Do we trim the "LNKxxxx" part of these messages somehow? Those are localization-independent.

**Comment by ChrisDenton** [comment]
> Do we trim the "LNKxxxx" part of these messages somehow? Those are localization-independent.

No, we should never trim any thing lines a warning/error code.

A suggestion I made before was hiding the whole output unless it contains at least one `LNKxxxx` code somewhere in there because all warnings and errors should have such a code. This would be better in the common case where there are no linker warnings. Then cleaning up that output could just be best effort (likely English only unless someone has other bright ideas).

**Comment by workingjubilee** [comment]
One of the things we filter has a LNK code apparently (LNK6004, I believe).

**Comment by ChrisDenton** [comment]
We could make an exception for LNK6004 but tbh I'm not sure why that needs filtering. We [always pass `/OPT:REF`](https://github.com/rust-lang/rust/blob/77cf889bc178ddb44d6a1c78e5a820b5abb31d8d/compiler/rustc_codegen_ssa/src/back/linker.rs#L1020-L1024) whether optimising or not, which is [documented](https://learn.microsoft.com/en-us/cpp/build/reference/incremental-link-incrementally?view=msvc-180) as being incompatible with incremental linking.

**Comment by workingjubilee** [comment]
> A suggestion I made before was hiding the whole output unless it contains at least one LNKxxxx code somewhere in there because all warnings and errors should have such a code. This would be better in the common case where there are no linker warnings. Then cleaning up that output could just be best effort (likely English only unless someone has other bright ideas).

My inclination is that we should ship this with no cleanup, and then do that after the fact if it becomes clear it is highly desirable.

**Comment by estebank** [comment]
Setting `VSLANG=1033` *if the english language pack is installed*, will make `link.exe` work in english. I found that `set DOTNET_CLI_UI_LANGUAGE=us-EN` is also considered by MSBuild in general (I'm assuming they have the same or similar behavior).

Once a warning has been constructed, we could invoke `msbuild.exe` to verify whether it returns an expected english string(s), so we could either extend the warning with a note mentioning that the developer should install the english pack, or outright silence the warning.
